Makefile.am

```diff
@@ -75,196 +84,208 @@ mon_srcs =                   \
     mon/mon_resp_decode.c    \
     mon/mon_resp_encode.c
 
-# monitoring server sources
+## monitoring server sources - must also include gss_srcs
 mons_srcs = \
     mon/mons.c \
     mon/mons_handlers.c
 
-check_PROGRAMS = common/t_constraint
-TESTS = common/t_constraint
+## Called by make check
+check_PROGRAMS = \
+    common/t_constraint \
+    tests/t_libtr_tid_la
+TESTS = \
+    common/t_constraint \
+    tests/t_libtr_tid_la
 TEST_CFLAGS = -Wno-missing-prototypes
 
 lib_LTLIBRARIES = libtr_tid.la
 
-libtr_tid_la_SOURCES = $(tid_srcs) \
-$(common_srcs) \
-trp/trp_req.c \
-trp/trp_upd.c
+libtr_tid_la_SOURCES = \
+    $(common_srcs) \
+    $(gss_srcs) \
+    $(tid_srcs) \
+    trp/trp_req.c \
+    trp/trp_upd.c
 
 libtr_tid_la_CFLAGS = $(AM_CFLAGS) -fvisibility=hidden
 libtr_tid_la_LIBADD = gsscon/libgsscon.la $(GLIB_LIBS)
 libtr_tid_la_LDFLAGS = $(AM_LDFLAGS) -version-info 4:3:2 -no-undefined
 
-common_t_constraint_SOURCES = common/t_constraint.c \
-common/tr_debug.c \
-common/tr_name.c \
-common/tr_list.c \
-common/tr_constraint.c \
-common/tr_dh.c \
-common/tr_rand_id.c \
-tid/tid_req.c \
-tid/tid_resp.c
+common_t_constraint_SOURCES = \
+    common/t_constraint.c \
+    common/tr_debug.c \
+    common/tr_name.c \
+    common/tr_list.c \
+    common/tr_constraint.c \
+    common/tr_dh.c \
+    common/tr_rand_id.c \
+    tid/tid_req.c \
+    tid/tid_resp.c
 
 common_t_constraint_CPPFLAGS = $(AM_CPPFLAGS) -DTESTS=\"$(srcdir)/common/tests.json\"
 common_t_constraint_LDADD = gsscon/libgsscon.la $(GLIB_LIBS)
 
-tr_trust_router_SOURCES =tr/tr_main.c \
-tr/tr.c \
-tr/tr_event.c \
-tr/tr_cfgwatch.c \
-tr/tr_tid.c \
-tr/tr_tid_mons.c \
-tr/tr_trp.c \
-tr/tr_trp_mons.c \
-tr/tr_mon.c \
-common/tr_gss.c \
-common/tr_gss_client.c \
-$(tid_srcs) \
-$(trp_srcs) \
-$(common_srcs) \
-$(mons_srcs)
+tests_t_libtr_tid_la_SOURCES = tests/t_libtr_tid_la.c
+tests_t_libtr_tid_la_LDADD = $(lib_LTLIBRARIES)
+
+tr_trust_router_SOURCES = \
+    $(common_srcs) \
+    tr/tr_main.c \
+    tr/tr.c \
+    tr/tr_event.c \
+    tr/tr_cfgwatch.c \
+    tr/tr_tid.c \
+    tr/tr_tid_mons.c \
+    tr/tr_trp.c \
+    tr/tr_trp_mons.c \
+    tr/tr_mon.c \
+    $(trp_srcs) \
+    $(gss_srcs) \
+    $(mons_srcs)
 
 tr_trust_router_LDFLAGS = $(AM_LDFLAGS) -levent_pthreads -pthread
-tr_trust_router_LDADD = gsscon/libgsscon.la $(GLIB_LIBS)
+tr_trust_router_LDADD = gsscon/libgsscon.la $(GLIB_LIBS) $(lib_LTLIBRARIES)
```

Comment:
If I'm reading this file correctly, it looks like `$(gss_srcs)` are among the `tr_trust_router_SOURCES` and also among the `libtr_tid_la_SOURCES`.  Since the `libtr_tid_la_SOURCES` files are compiled into `lib_LTLIBRARIES`, wouldn't that make their inclusion in `tr_trust_router_SOURCES` redundant?

Reply:
That is correct. However, the contents of `$(gss_srcs)` are not part of the public interface to the shared  library - they're used by the library code itself, but not available to an external binary.

I don't think we want to commit to supporting that as part of the ABI.

This arrangement allows the trust_router package to be of a different version than the libtr_tid package (these are separate on Debian). This introduces the possibility of incompatibilities and means we need to test various combinations, but I think it's the right thing to do. If the trust_router cannot operate against all the versions of the library we declare to be ABI-compatible, that's a bug in the libraries that needs to be caught.

Reply:
That sounds sane.